cpp/src/arrow/compute/kernels/codegen_internal.h

```diff
@@ -787,6 +830,41 @@ ArrayKernelExec Integer(detail::GetTypeId get_id) {
   }
 }
 
+template <template <typename...> class Generator, typename Type0, typename... Args>
+ArrayKernelExec IntegerBased(detail::GetTypeId get_id) {
```

Comment:
What's this? How is it supposed to be used? `codegen_internal.h` quickly seems to become write-only: it's difficult to make sense of why those things exist and what they're good at.

Reply:
I'll document it, sorry.

Reply:
FWIW, we have slight precedent for naming this "IsPhysicalInteger": `type_traits.h::is_physical_integer_type`